[FLINK-13984] Separate on-heap and off-heap managed memory pools

azagrebin opened the pull request:
## What is the purpose of the change

This does some refactoring to prepare and then changes the MemoryManager to serve both types of memory at the same time: HEAP and OFF_HEAP. Additionally to paged segments, it adds API to reserve arbitrary sized chunks of memory.

## Brief change log

  - some cleanup in MemoryManager
  - add MemoryManagerBuilder for tests
  - deprecate memory preallocation in MemoryManager
  - implement KeyedBudget to atomically track available memory per type up to the limit.
  - change MemoryManager to use KeyedBudget instead of internal locking
  - switch to concurrent hash maps for owner tracking in MemoryManager
  - add another type of memory to MemoryManager
  - Implement reservation of memory chunks in MemoryManager

## Verifying this change

Existing and newly added unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented:
cc @xintongsong

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -257,133 +216,98 @@ public boolean verifyEmpty() {
-				segmentsForOwner = new HashSet<MemorySegment>(numPages);
-				allocatedSegments.put(owner, segmentsForOwner);
-			}
+		// in the case of pre-allocated memory, the 'numNonAllocatedPages' is zero, in the
```
Good catch, thanks, I will change it.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -394,74 +318,79 @@ public void release(Collection<MemorySegment> segments) {
 					}
+				}
+				while (segment != null) {
+					segment = releaseSegmentsForOwnerUntilNextOwner(segment, segmentsIterator, releasedMemory);
```
True, it would simplify the things. I thought about it. The original code was doing only one traversal over the segments to release, I think to optimise the code and spare memory. I decided to keep this approach to avoid any possible degradation. Although, we can later try whether there is actual performance degradation and change it to the simpler approach for better maintainability.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -474,60 +403,121 @@ public void releaseAll(Object owner) {
 	 */
-	public MemoryType getMemoryType() {
-		return memoryType;
+	public void releaseMemory(Object owner, MemoryType memoryType, long size) {
```
We could, but I cannot imagine any use case at the moment. For segments, the memory type does not matter but for chunks, there is usually an intention which type to use.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -474,60 +403,121 @@ public void releaseAll(Object owner) {
+				reservations.compute(
+					memoryType,
+					(mt, currentlyReserved) ->
+						currentlyReserved == null || currentlyReserved <= size ? null : currentlyReserved - size);
```
Good point, we did not have it previously for segments (maybe because double-release is acceptable for some use cases in Dataset API atm) but it can be useful for debugging, I will add it.

azagrebin commented:
Thanks for the review @KarmaGYZ 
I have updated the PR.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -257,133 +216,96 @@ public boolean verifyEmpty() {
 	 */
-	public void allocatePages(Object owner, List<MemorySegment> target, int numPages)
-			throws MemoryAllocationException {
+	public void allocatePages(
```
In addition to this, I think we also need 'allocatePages' that support specifying memory types.

The current 'allocatePages', which do not specify which type (on-heap / off-heap) of memory should be allocated, may work well with DataSet jobs without fine grained resource requirement. But for blink sql jobs (bounded streaming) with fine grained resource profiles (flip-53), operators have specific and separate memory quota for on-heap / off-heap managed memory,  and will be scheduled accordingly. Thus even they can use pages of either type, they need to allocate pages respecting their quota for each pool. Otherwise, other operators that can only use one of the type (e.g., flip-58 python udx that uses off-heap managed memory) might be affected.

azagrebin replied:
Alright, if there are already known use cases with the mixed allocation type, then we should add those methods.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -568,144 +569,26 @@ public int computeNumberOfPages(double fraction) {
+		//noinspection NumericCastThatLosesPrecision
 		return (int) (totalNumPages * fraction / numberOfSlots);
 	}
 
```
I think we also need 'computeNumberOfPages' for specific memory type, for the same reason as 'allocatePages'.

azagrebin replied:
I still think that the fraction concern is not strictly related to the memory management per se but it is rather a higher level concern about how the memory is distributed among the operators. I think the main reason we had this method was even to hold the number of slots which should go away when the memory manager is per slot.

Fractions and the memory limits can be handled in one place where the operators get the memory manager and their limits which they can decide how to break into chunks and pages.
Anyways, I agree we can reiterate on this later during the MM per slot refactoring.

Instead, I suggest we expose the total number of pages per type at the moment, same as memory size per type which is naturally related to the memory management.

xintongsong replied:
I think the reason we had this method was not only hold the number of slots, but also hold the total number of pages. In other words, MM knows how many memory in total the slot has, which will not change when we have MM per slot.

I agree that computing operator level memory budget from the slot total memory and the operator fraction is another responsibility for MM. It makes sense that MM exposes the slot total memory and operator calls a util to compute the budget from the slot total memory and the fraction.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -564,25 +529,6 @@ public int computeNumberOfPages(double fraction) {
-	 * @param fraction The fraction of the memory of the task slot.
-	 * @return The number of pages corresponding to the memory fraction.
-	 */
-	public long computeMemorySize(double fraction) {
```
True, we could leave this method for now. I am a bit sceptical about these methods because they add another concern to this class which does not look like being directly related to the manager because it does not do anything with fractions. I would prefer to do it on the user side by calling `getMemorySize/NumberOfPages` etc methods or introduce a utility method to deduplicate the code.

My plan was that if we make memory manager per slot then the operator limit calculation is even more straightforward and I would refactor it afterwards.

xintongsong replied:
I think it makes sense to keep the two methods in memory manger, so that only memory manager but not operators needs to known the total managed memory size of the slot. Let's leave the discussion to when we make memory manager per slot.

azagrebin commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -234,16 +234,15 @@
 	/**
 	 * Whether TaskManager managed memory should be pre-allocated when the TaskManager is starting.
 	 */
+	@Deprecated
```
True, if it is configured it will be just ignored. I wanted to leave it for the documentation purposes but the better place is release notes. I will remove it.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -300,31 +255,21 @@ public void allocatePages(Object owner, List<MemorySegment> target, int numPages
 	 *
-	 * <p>If the memory manager manages pre-allocated memory, the memory segment goes back to the memory pool.
-	 * Otherwise, the segment is only freed and made eligible for reclamation by the GC.
+	 * <p>If the segment has already been released or is null, the request is simply ignored.
```
Remove completely?

I can change it to:
`If the segment has already been released, it is only freed. If it is null or has no owner, the request is simply ignored.`

Also, two more things came into my mind, we should not release budget if there is no such segment registered for this owner (e.g. double release) and we should always try to remove the segment even if it is freed because it might have been freed outside but not released.

xintongsong replied:
My bad, please ignore this comment.

I think I found some kind of typo with an extra 'if' in the doc/comment when I reviewed this PR locally in my IDE. But I cannot find it now. I probably have left this comment at wrong place when moving from IDE to GitHub.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -288,10 +244,9 @@ public void allocatePages(Object owner, List<MemorySegment> target, int numPages
 
 			// in the case of pre-allocated memory, the 'numNonAllocatedPages' is zero, in the
 			// lazy case, the 'freeSegments.size()' is zero.
```
True, it should be just removed

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/KeyedBudget.java:
```diff
@@ -0,0 +1,169 @@
+import java.util.Map;
+import java.util.Map.Entry;
+
+class KeyedBudget<K> {
```
ok, I will add a javadoc with a general class description. For class name, not sure about the better name, I could name it `KeyedBudgetManager`.

Regarding the `MemorySize` usage, `KeyedBudget` is designed as a general purpose component, it is not necessary about memory. Maybe, it is even better idea if I move it to the `util` package.

True, `MemorySize` is a nicer API. `MemorySize` is also a mostly user facing class from what I see. My thinking is that `MemoryManger` is a Flink internal component and all its users does not seem to be interested in memory units atm but rather all work with bytes. Although wrapping will not add too much overhead, I would keep it as it is now. We can always do a separate follow-up hotfix where we add/switch to `MemorySize` API if there will be use cases for memory units in `MemoryManger`.

xintongsong replied:
I cannot come up with a better descriptive name neither. `KeyedBudgetManager` sounds better than `KeyedBudget`, but still not as intuitional as that people can immediately tell what does this class do from its name. I guess sometimes it's not easy to always find super matching names for a classes, and in those cases having some docs explaining the general purpose and responsibility of the class should be good enough.

Regarding `MemorySize` vs. bytes in long, I'm not sure how general this `KeyedBudget` could be used. But agree that using `MemorySize` is not a critical need, and if needed we can do it easily with a follow-up. FTM, let's keep it the current way.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/memory/KeyedBudgetTest.java:
```diff
@@ -0,0 +1,154 @@
+		long acquired = keyedBudget.acquireBudgetForKey("k1", 20L);
+
+		assertThat(acquired, is(15L));
+		checkOneKeyBudgetChange(keyedBudget, "k1", 15L);
```
True, it will be a bit more readable

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/memory/KeyedBudgetTest.java:
```diff
@@ -0,0 +1,154 @@
+		assertThat(keyedBudget.availableBudgetForKeys(Arrays.asList("k1", "k4")), is(8L));
+		assertThat(keyedBudget.totalAvailableBudget(), is(32L));
+	}
+
```
`acquirePagedBudget` is just a straightforward shorthand of `acquirePagedBudgetForKeys`, I think we do not need to basically duplicate all tests for it.

max possible budget is tested in the `assertThat` inside `testFailedAcquisitionForKey`.

Agree, the failure test for `acquirePagedBudgetForKeys` is forgotten. I will add it.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -402,29 +386,25 @@ public void releaseAll(Object owner) {
+		// free each segment
+		EnumMap<MemoryType, Long> releasedMemory = new EnumMap<>(MemoryType.class);
+		for (MemorySegment segment : segments) {
+			segment.free();
```
not sure about what is meant in this comment, maybe it was done before some changes.
Currently, `releaseSegment` does not do `segment.free()`.

xintongsong replied:
True, this comment is out of sync.

azagrebin commented:
Thanks for the review @xintongsong 
I addressed the comments

KarmaGYZ commented:
Thanks for addressing the comments. 
I have reviewed latest version. +1 for merging.

azagrebin commented:
Thanks for the reviews @xintongsong @KarmaGYZ 
I removed the pre-allocation left-overs, I guess got lost during rebases, thx

carp84 commented on flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java:
```diff
@@ -474,67 +424,141 @@ public void releaseAll(Object owner) {
+			return;
+		}
+
+		reservedMemory.compute(owner, (o, reservations) -> {
```
It seems we're using a "fail silently" way here if the `compute` result is `null` which means we cannot find the owner in the map anymore. Is it possible that one releases its reserved memory multiple times? If so, does it worth any logging here?

azagrebin replied:
As I understand, we already have a log message for this:
```
LOG.warn(
	"Trying to release more memory {} than it was reserved {} so far for the owner {}",
	size,
	currentlyReserved == null ? 0 : currentlyReserved,
	owner);
```
or do you mean something else?
or the older code was reviewed :)

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java:
```diff
@@ -0,0 +1,238 @@
+
+	public void releasePageForKey(K key) {
+		releaseBudgetForKey(key, defaultPageSize);
+	}
```
True, I could add a test. The reason, I have not added, is that this is just a shorthand method as also others. The value of this kind of tests is probably not so big comparing to the effort and maintaining them. It is also an internal component, in worst un-probable case, tests for other components using them will fail.

carp84 replied:
True for the existing implementation, but things may change if the implementation changes later, like what would happen if my suggestion on narrow down the scope of the `lock` makes sense.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java:
```diff
@@ -0,0 +1,238 @@
+		synchronized (lock) {
+			availableBudgetByKey.putAll(maxBudgetByKey);
+		}
+	}
```
I could add check after each test that after `releaseAll`, all budget is available.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/KeyedBudgetManagerTest.java:
```diff
@@ -0,0 +1,178 @@
+ * Test suite for {@link KeyedBudgetManager}.
+ */
+@SuppressWarnings("MagicNumber")
+public class KeyedBudgetManagerTest {
```
The methods are simply wrapped by global lock at the moment, basically serialised as if into one thread. What kind of concurrent tests do you think we miss?

carp84 replied:
I mean if we try to call `acquireBudgetForKey` or `acquirePagedBudgetForKeys` in multiple threads, the counting of budgets won't be violated. Yes the logic is assured by the implementation but we need some test case to cover it and prevent future changes follow the same correctness. Wdyt?

azagrebin replied:
I added `KeyedBudgetManagerTest.testConcurrentAcquisitionForKeys` and `KeyedBudgetManagerTest.testConcurrentReleaseForKeys`.

carp84 commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java:
```diff
@@ -0,0 +1,238 @@
+	 * @return the acquired number of pages for each key if the acquisition is successful (either left) or
+	 * the total number of pages which were available for the given keys (either right).
+	 */
+	Either<Map<K, Long>, Long> acquirePagedBudgetForKeys(Iterable<K> keys, long numberOfPages, long pageSize) {
```
Yes `BudgetAllocationException` would be more accurate, and my suggestion comes from the fact that now `MemoryManager.reserveMemory` also throws a `MemoryAllocationException`, and I agree change the method signature could better reflect the case.

Personally I think the semantic is explicit to return the acquired budget when succeed and throw exception when fail. Having an internal functional `Try` allows incorrect handling in upper caller, like if `MemoryManager.reserveMemory` doesn't check the returning value and continues, no warning/error will happen but there would be memory leak. Throwing an explicit exception will also make the test coverage tool working better (thus help easier maintenance), for the existing implementation I need to manually verify whether `Either.right` is checked in test case since it won't be reflected by test coverage.

azagrebin replied:
Ok, I added `AllocationResult` which throws as exception if result is not checked but the wrong is queried.

carp84 commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java:
```diff
@@ -0,0 +1,238 @@
+	@GuardedBy("lock")
+	private final Map<K, Long> availableBudgetByKey;
+
+	private final Object lock = new Object();
```
My suggestion is on `acquireBudgetForKey`, not `acquirePagedBudgetForKeys`. I could see you're considering the concurrent call on `MemoryManager.reserveMemory` based on the fact that `MemoryManager.reservedMemory` is a `ConcurrentHashMap`, and the current implementation will make such effort useless.

To be more explicit, I think the big lock is designed for "transaction" of acquiring budget from multiple keys, and we could have better handling to avoid contention for the non-transactional acquirement on single key, like acquirement for on-heap and off-heap pages could happen in parallel rather than blocking each other.

azagrebin replied:
Offline conclusion: change the implementation to `ConcurrentHashMap` as a follow up when needed.

azagrebin commented:
Thanks for the review @carp84
I addressed the comments and introduced `MemoryReservationException` for `MemoryManager.reserveMemory` instead of `MemoryAllocationException`.

azagrebin commented:
Thanks @carp84 , merging it when CI is green
